src/map/battle.cpp

```diff
@@ -938,62 +938,216 @@ static void battle_absorb_damage(struct block_list *bl, struct Damage *d) {
 }
 
 /**
- * Check Safety Wall and Pneuma effect.
- * Maybe expand this to move checks the target's SC from battle_calc_damage?
+ * Check for active statuses that block damage
  * @param src Attacker
  * @param target Target of attack
  * @param sc STatus Change
  * @param d Damage data
  * @param damage Damage received
  * @param skill_id
  * @param skill_lv
- * @return True:Damage inflicted, False:Missed
+ * @return True: Damage inflicted, False: Missed
  **/
 bool battle_check_sc(struct block_list *src, struct block_list *target, struct status_change *sc, struct Damage *d, int64 damage, uint16 skill_id, uint16 skill_lv) {
+	struct map_session_data *sd = NULL;
+	struct status_change_entry *sce;
+	int flag = d->flag;
+
 	if (!sc)
 		return true;
 
-	if (sc->data[SC_SAFETYWALL] && (d->flag&(BF_SHORT|BF_MAGIC)) == BF_SHORT) {
-		struct skill_unit_group* group = skill_id2group(sc->data[SC_SAFETYWALL]->val3);
-		uint16 skill_id_val = sc->data[SC_SAFETYWALL]->val2;
+	if (target->type == BL_PC)
+		sd = (struct map_session_data *)target;
 
-		if (group) {
-			if (skill_id_val == MH_STEINWAND) {
-				if (--group->val2 <= 0)
-					skill_delunitgroup(group);
-				d->dmg_lv = ATK_BLOCK;
-				if( (group->val3 - damage) > 0 )
-					group->val3 -= (int)cap_value(damage, INT_MIN, INT_MAX);
-				else
-					skill_delunitgroup(group);
-				return false;
-			}
-			//in RE, SW possesses a lifetime equal to group val2, (3x caster hp, or homon formula)
+	if (sc->data[SC_PNEUMA] && (d->flag&(BF_MAGIC | BF_LONG)) == BF_LONG ||
+		sc->data[SC_BASILICA] && !status_bl_has_mode(src, MD_STATUS_IMMUNE) ||
+		sc->data[SC_ZEPHYR] && !(flag&BF_MAGIC && skill_id) && !(skill_get_inf(skill_id)&(INF_GROUND_SKILL | INF_SELF_SKILL)) ||
+		sc->data[SC__MANHOLE] ||
+		sc->data[SC_KINGS_GRACE]
+		)
```

Comment:
I don't see parentheses to covers those `&&`s
 `(pneuma && range) || (basilica && immune) || ...`